timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py

```diff
@@ -400,13 +404,15 @@ def slice_by_timestep(self, time_step_slice: slice) -> TimeSeriesDataFrame:
             Data frame containing only the time steps of each ``item_id`` sliced according to the
             input ``time_step_slice``.
         """
-        slice_gen = ((i, self.loc[i].iloc[time_step_slice]) for i in self._item_index)
-        slices = []
-        for ix, data_slice in slice_gen:
-            idx = pd.MultiIndex.from_product([(ix,), data_slice.index], names=[ITEMID, TIMESTAMP])
-            data_slice.set_index(idx, inplace=True)
-            slices.append(data_slice)
-        return self.__class__(pd.concat(slices), static_features=self.static_features)
+        num_timesteps_per_item = self.num_timesteps_per_item()
+        # Create a boolean index that selects the correct slice in each timeseries
+        boolean_indicators = []
+        for length in num_timesteps_per_item:
+            indicator = np.zeros(length, dtype=bool)
+            indicator[time_step_slice] = True
+            boolean_indicators.append(indicator)
+        index = np.concatenate(boolean_indicators)
+        return self.__class__(self.loc[index].copy(), static_features=self.static_features)
```

Comment:
nit: it's very surprising `loc` works here and it does feel a little wobbly. I'd leave it in for now, but it felt a little too much of overloading on pandas part.

Reply:
Changed this to `self[index].copy()`, I agree that this makes more sense